src/FNA3D_Tracing.c

```diff
@@ -170,26 +170,41 @@ void FNA3D_Trace_RegisterEffect(FNA3D_Effect *effect, MOJOSHADER_effect *effectD
 }
 #undef TRACE_OBJECT
 
-#define WRITE(val) ops->write(ops, &val, sizeof(val), 1)
+#define CHECK_AND_FLUSH_BUFFER(len) if (traceBufferCurrentSize + len > traceBufferSize) { FNA3D_Trace_FlushMemory(); }
+#define WRITE(val) CHECK_AND_FLUSH_BUFFER(sizeof(val)); SDL_memcpy((uint8_t*)traceBuffer + traceBufferCurrentSize, &val, sizeof(val)); traceBufferCurrentSize += sizeof(val);
+#define WRITEMEM(ptr, len) CHECK_AND_FLUSH_BUFFER(len); SDL_memcpy((uint8_t*)traceBuffer + traceBufferCurrentSize, ptr, len); traceBufferCurrentSize += len;
```

Comment:
We can probably expand these now...

```suggestion
#define CHECK_AND_FLUSH_BUFFER(len) \
	if (traceBufferCurrentSize + len > traceBufferSize) \
	{ \
		FNA3D_Trace_FlushMemory(); \
	}
#define WRITE(val) \
	CHECK_AND_FLUSH_BUFFER(sizeof(val)) \
	SDL_memcpy((uint8_t*) traceBuffer + traceBufferCurrentSize, &val, sizeof(val)); \
	traceBufferCurrentSize += sizeof(val);
#define WRITEMEM(ptr, len) \
	CHECK_AND_FLUSH_BUFFER(len) \
	SDL_memcpy((uint8_t*) traceBuffer + traceBufferCurrentSize, ptr, len); \
	traceBufferCurrentSize += len;
```